### PR TITLE
[WIP] Support DiLoCo training

### PR DESCRIPTION
**Pull request by jonb377**
DiLoCo training paper: https://arxiv.org/pdf/2311.08105

Adds preliminary support for DiLoCo training through DrJax.

### CODE REVIEW

File: MaxText/diloco.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    avg = jax.tree.map(lambda x: (x / config.num_diloco_replicas).astype(x.dtype), total)
+    return avg
+
+  @drjax.program(placements={'diloco': config.num_diloco_replicas})
+  def diloco_train_step(state, batch, prng):
+    # Batch shape is [pdbs, S]. To use in DiLoCo, we need to add a leading axis
+    # to vmap over.
+    per_replica_batch = config.global_batch_size_to_train_on // config.num_diloco_replicas
+    batch_shape = (config.num_diloco_replicas, per_replica_batch, -1)
+    batch = jax.tree.map(lambda x: x.reshape(batch_shape), batch)
```
Comment (zcharles8) [inline MaxText/diloco.py:103]: In some versions I've written, I explicitly add this extra axis before passing to the train step, so that I can ensure that it is sharded over the diloco axis. I'm not certain if that's necessary though, but wanted to flag it for posterity.

File: MaxText/diloco.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    # Broadcast the RNG across replicas
+    broadcast_rng = drjax.broadcast(prng)
+
+    inner_state, metrics = drjax.map_fn(
+      train_step,
+      (state.inner_state, batch, broadcast_rng)
+    )
+    avg_metrics = typed_reduce_mean(metrics)
```
Comment (zcharles8) [inline MaxText/diloco.py:112]: One thing to call out is that differentiating between metrics that need to be summed and metrics that need to be averaged is kind of a pain.

E.g. if you're keeping track of total number of tokens seen, then a reduce_sum here is probably desired. I don't have any great solutions though without understanding better the structure of metrics.
